examples/hmcecs.py

```diff
@@ -0,0 +1,147 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+"""
+Example: Hamiltonian Monte Carlo with Energy Conserving Subsampling
+=========================
+
+This example illustrates the use of subsampling of HMC using Energy Conserving Subsampling when the likelihood is
+additive.
+
+**References:**
+
+    1. *Hamiltonian Monte Carlo with energy conserving subsampling*,
+       Dang, K. D., Quiroz, M., Kohn, R., Minh-Ngoc, T., & Villani, M. (2019)
+
+    :align: center
+"""
+
+import argparse
+import time
+
+import matplotlib.pyplot as plt
+
+from jax import random
+import jax.numpy as jnp
+
+import numpyro
+import numpyro.distributions as dist
+from numpyro.examples.datasets import HIGGS, load_dataset
+from numpyro.infer import HMC, HMCECS, MCMC, NUTS, SVI, Trace_ELBO, autoguide
+from numpyro.infer.hmc_gibbs import taylor_proxy
+
+
+def model(data, obs, subsample_size):
+    n, m = data.shape
+    theta = numpyro.sample('theta', dist.continuous.Normal(jnp.zeros(m), .5 * jnp.ones(m)))
+    with numpyro.plate('N', n, subsample_size=subsample_size):
+        batch_feats = numpyro.subsample(data, event_dim=1)
+        batch_obs = numpyro.subsample(obs, event_dim=0)
+        numpyro.sample('obs', dist.Bernoulli(logits=theta @ batch_feats.T), obs=batch_obs)
+
+
+def run_hmcecs(hmcecs_key, args, data, obs, inner_kernel):
+    svi_key, mcmc_key = random.split(hmcecs_key)
+
+    # find reference parameterization
+    optimizer = numpyro.optim.Adam(step_size=1e-3)
+    guide = autoguide.AutoDelta(model)
+    svi = SVI(model, guide, optimizer, loss=Trace_ELBO())
+    params, losses = svi.run(svi_key, args.num_svi_steps, data, obs, args.subsample_size)
+    ref_params = {'theta': params['theta_auto_loc']}
+
+    # setup proxy
+    proxy = taylor_proxy(ref_params)
+
+    kernel = HMCECS(inner_kernel, num_blocks=args.num_blocks, proxy=proxy)
+    mcmc = MCMC(kernel, num_warmup=args.num_warmup, num_samples=args.num_samples)
+
+    mcmc.run(mcmc_key, data, obs, args.subsample_size)
+    mcmc.print_summary()
+    return losses, mcmc.get_samples()
+
+
+def run_hmc(mcmc_key, args, data, obs, kernel):
+    mcmc = MCMC(kernel, num_warmup=args.num_warmup, num_samples=args.num_samples)
+    mcmc.run(mcmc_key, data, obs, None)
+    mcmc.print_summary()
+    return mcmc.get_samples()
+
+
+def main(args):
+    assert 11_000_000 >= args.num_datapoints, "11,000,000 data points in the Higgs dataset"
+    # full dataset takes hours for plain hmc!
+    _, fetch = load_dataset(HIGGS, shuffle=False, num_datapoints=args.num_datapoints)
+    data, obs = fetch()
+
+    hmcecs_key, hmc_key = random.split(random.PRNGKey(args.rng_seed))
+
+    # choose inner_kernel
+    if args.inner_kernel.lower() == 'hmc':
+        inner_kernel = HMC(model)
+    elif args.inner_kernel.lower() == 'nuts':
+        inner_kernel = NUTS(model)
+    else:
```

Comment:
get rid of this?

Reply:
If `inner_kernel` is `None`, HMCECS becomes naive subsampling which Betancourt showed to samples from a bias trajectory, but the option is in NumPyro, which is why I figured it would be worth displaying. I could change the choice to `naive` instead.

Reply:
gotcha yeah just add `naive` to `choices` 👍